File: apps/api/plane/db/models/contract.py
# Python imports
from django.db import models
from django.conf import settings

# Package imports
from .base import BaseModel


# Contract type choices for government contracts
CONTRACT_TYPE_CHOICES = [
    ("idiq", "IDIQ"),  # Indefinite Delivery/Indefinite Quantity
    ("bpa", "BPA"),  # Blanket Purchase Agreement
    ("ffp", "Firm Fixed Price"),
    ("tm", "Time & Materials"),
    ("cpff", "Cost Plus Fixed Fee"),
    ("task_order", "Task Order"),  # Under IDIQ
    ("call_order", "Call Order"),  # Under BPA
]

CONTRACT_HIERARCHY_CHOICES = [
    ("master", "Master Contract"),  # IDIQ or BPA umbrella
    ("task_order", "Task Order"),  # Work order under IDIQ
    ("call_order", "Call Order"),  # Order under BPA
    ("standalone", "Standalone"),  # FFP, T&M, CPFF not under umbrella
]

CONTRACT_STATUS_CHOICES = [
    ("active", "Active"),
    ("completed", "Completed"),
    ("expired", "Expired"),
    ("cancelled", "Cancelled"),
]


class Contract(BaseModel):
    """
    Government contract model supporting IDIQ, BPA, FFP, T&M, and CPFF contract types.

    All monetary amounts are stored in cents (integer) to avoid floating-point errors.
    For example, $100,000 is stored as 10000000 cents.
    """

    workspace = models.ForeignKey(
        "db.Workspace",
        on_delete=models.CASCADE,
        related_name="contracts",
    )

    # Basic contract info
    name = models.CharField(max_length=255)
    contract_number = models.CharField(max_length=100, blank=True)
    description = models.TextField(blank=True)

    # Contract type and hierarchy
    contract_type = models.CharField(
        max_length=20,
        choices=CONTRACT_TYPE_CHOICES,
        default="ffp",
    )
    hierarchy_type = models.CharField(
        max_length=20,
        choices=CONTRACT_HIERARCHY_CHOICES,
        default="standalone",
    )
    parent_contract = models.ForeignKey(
        "self",
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name="child_contracts",
        help_text="Parent IDIQ/BPA contract for task orders and call orders",
    )

    # Vendor/Contractor info
    vendor_name = models.CharField(max_length=255, blank=True)
    vendor_contact = models.CharField(max_length=255, blank=True)

    # Financial amounts (stored in cents)
    ceiling_amount = models.BigIntegerField(
        default=0,
        help_text="Maximum contract value in cents (e.g., 10000000 = $100,000)",
    )
    obligated_amount = models.BigIntegerField(
        default=0,
        help_text="Committed/obligated funds in cents",
    )
    expended_amount = models.BigIntegerField(
        default=0,
        help_text="Actual spend to date in cents",
    )
    currency = models.CharField(max_length=3, default="USD")

    # Contract dates
    start_date = models.DateField(null=True, blank=True)
    end_date = models.DateField(null=True, blank=True)
    pop_start = models.DateField(
        null=True,
        blank=True,
        help_text="Period of Performance start date",
    )
    pop_end = models.DateField(
        null=True,
        blank=True,
        help_text="Period of Performance end date",
    )

    # Status
    status = models.CharField(
        max_length=20,
        choices=CONTRACT_STATUS_CHOICES,
        default="active",
    )

    # Flexible metadata for additional contract-specific data
    metadata = models.JSONField(default=dict, blank=True)

    # Sort order for manual ordering
    sort_order = models.FloatField(default=65535)

    class Meta:
        verbose_name = "Contract"
        verbose_name_plural = "Contracts"
        db_table = "contracts"
        ordering = ("-created_at",)
        unique_together = [["workspace", "contract_number", "deleted_at"]]
        constraints = [
            models.UniqueConstraint(
                fields=["workspace", "contract_number"],
                condition=models.Q(deleted_at__isnull=True) & ~models.Q(contract_number=""),
                name="contract_unique_workspace_number_when_not_deleted",
            )
        ]

    def __str__(self):
        if self.contract_number:
            return f"{self.contract_number} - {self.name}"
        return self.name

    @property
    def remaining_amount(self):
        """Calculate remaining budget (ceiling - obligated)."""
        return self.ceiling_amount - self.obligated_amount

    @property
    def available_amount(self):
        """Calculate available funds (obligated - expended)."""
        return self.obligated_amount - self.expended_amount

    @property
    def utilization_percentage(self):
        """Calculate contract utilization percentage."""
        if self.ceiling_amount == 0:
            return 0
        return (self.obligated_amount / self.ceiling_amount) * 100


class ContractProjectAllocation(BaseModel):
    """
    Junction table linking contracts to projects with allocation amounts.

    A contract can fund multiple projects, and a project can be funded by multiple contracts.
    This table tracks how much of each contract is allocated to each project.
    """

    contract = models.ForeignKey(
        Contract,
        on_delete=models.CASCADE,
        related_name="project_allocations",
    )
    project = models.ForeignKey(
        "db.Project",
        on_delete=models.CASCADE,
        related_name="contract_allocations",
    )

    # Financial allocation (stored in cents)
    allocated_amount = models.BigIntegerField(
        default=0,
        help_text="Amount allocated from this contract to this project in cents",
    )
    actual_spend = models.BigIntegerField(
        default=0,
        help_text="Actual spend against this allocation in cents",
    )

    # Optional fiscal year tracking (government contracts often have FY constraints)
    fiscal_year = models.CharField(max_length=10, blank=True)

    # Description of what this allocation covers
    description = models.TextField(blank=True)

    class Meta:
        verbose_name = "Contract Project Allocation"
        verbose_name_plural = "Contract Project Allocations"
        db_table = "contract_project_allocations"
        ordering = ("-created_at",)
        # Allow multiple allocations per contract-project pair (e.g., different fiscal years)
        # but prevent exact duplicates
        unique_together = [["contract", "project", "fiscal_year", "deleted_at"]]
        constraints = [
            models.UniqueConstraint(
                fields=["contract", "project", "fiscal_year"],
                condition=models.Q(deleted_at__isnull=True),
                name="allocation_unique_contract_project_fy_when_not_deleted",
            )
        ]

    def __str__(self):
        return f"{self.contract.name} → {self.project.name}"

    @property
    def remaining_allocation(self):
        """Calculate remaining allocation (allocated - actual_spend)."""
        return self.allocated_amount - self.actual_spend

    def save(self, *args, **kwargs):
        # Ensure contract and project are in the same workspace
        if self.contract.workspace_id != self.project.workspace_id:
            raise ValueError("Contract and project must be in the same workspace")
        super().save(*args, **kwargs)

File: packages/types/src/contract.ts
// Contract Types for Government Resource Management

// Contract type choices matching Django model
export type TContractType =
  | "idiq" // Indefinite Delivery/Indefinite Quantity
  | "bpa" // Blanket Purchase Agreement
  | "ffp" // Firm Fixed Price
  | "tm" // Time & Materials
  | "cpff" // Cost Plus Fixed Fee
  | "task_order" // Task Order (under IDIQ)
  | "call_order"; // Call Order (under BPA)

// Contract hierarchy choices
export type TContractHierarchyType =
  | "master" // Master Contract (IDIQ or BPA umbrella)
  | "task_order" // Work order under IDIQ
  | "call_order" // Order under BPA
  | "standalone"; // FFP, T&M, CPFF not under umbrella

// Contract status choices
export type TContractStatus = "active" | "completed" | "expired" | "cancelled";

// Main Contract type
export type TContract = {
  id: string;
  workspace: string;
  name: string;
  contract_number: string;
  description: string;
  contract_type: TContractType;
  hierarchy_type: TContractHierarchyType;
  parent_contract: string | null;
  vendor_name: string;
  vendor_contact: string;
  // All amounts stored in cents (integer)
  ceiling_amount: number;
  obligated_amount: number;
  expended_amount: number;
  currency: string;
  start_date: string | null;
  end_date: string | null;
  pop_start: string | null;
  pop_end: string | null;
  status: TContractStatus;
  metadata: Record<string, unknown>;
  sort_order: number;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  updated_by: string | null;
  // Computed fields from serializer
  remaining_amount?: number;
  available_amount?: number;
  utilization_percentage?: number;
  child_contracts_count?: number;
  project_allocations_count?: number;
};

// Lite version for lists and references
export type TContractLite = Pick<
  TContract,
  | "id"
  | "name"
  | "contract_number"
  | "contract_type"
  | "hierarchy_type"
  | "status"
  | "ceiling_amount"
  | "obligated_amount"
  | "expended_amount"
  | "start_date"
  | "end_date"
>;

// Form data for creating/updating contracts
export type TContractFormData = {
  name: string;
  contract_number?: string;
  description?: string;
  contract_type?: TContractType;
  hierarchy_type?: TContractHierarchyType;
  parent_contract?: string | null;
  vendor_name?: string;
  vendor_contact?: string;
  ceiling_amount?: number;
  obligated_amount?: number;
  expended_amount?: number;
  currency?: string;
  start_date?: string | null;
  end_date?: string | null;
  pop_start?: string | null;
  pop_end?: string | null;
  status?: TContractStatus;
  metadata?: Record<string, unknown>;
};

// Contract-Project Allocation types
export type TContractProjectAllocation = {
  id: string;
  contract: string;
  project: string;
  allocated_amount: number;
  actual_spend: number;
  fiscal_year: string;
  description: string;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  updated_by: string | null;
  // Computed fields
  remaining_allocation?: number;
  // Expanded fields
  contract_detail?: TContractLite;
};

// Form data for creating/updating allocations
export type TContractProjectAllocationFormData = {
  project: string;
  allocated_amount: number;
  actual_spend?: number;
  fiscal_year?: string;
  description?: string;
};

// Filter options for contract list
export type TContractFilters = {
  status?: TContractStatus;
  contract_type?: TContractType;
  hierarchy_type?: TContractHierarchyType;
  parent_contract?: string;
};

// Summary stats for dashboard
export type TContractSummary = {
  total_contracts: number;
  total_ceiling: number;
  total_obligated: number;
  total_expended: number;
  by_status: Record<TContractStatus, number>;
  by_type: Record<TContractType, number>;
};

// Display labels for contract types
export const CONTRACT_TYPE_LABELS: Record<TContractType, string> = {
  idiq: "IDIQ",
  bpa: "BPA",
  ffp: "Firm Fixed Price",
  tm: "Time & Materials",
  cpff: "Cost Plus Fixed Fee",
  task_order: "Task Order",
  call_order: "Call Order",
};

// Display labels for hierarchy types
export const CONTRACT_HIERARCHY_LABELS: Record<TContractHierarchyType, string> = {
  master: "Master Contract",
  task_order: "Task Order",
  call_order: "Call Order",
  standalone: "Standalone",
};

// Display labels for status
export const CONTRACT_STATUS_LABELS: Record<TContractStatus, string> = {
  active: "Active",
  completed: "Completed",
  expired: "Expired",
  cancelled: "Cancelled",
};
